File: services/gateway/src/headers.ts
import { getHeader } from "./auth";
import type { HeaderMap } from "./types";

const HOP_BY_HOP = new Set([
  "connection",
  "keep-alive",
  "proxy-authenticate",
  "proxy-authorization",
  "te",
  "trailer",
  "transfer-encoding",
  "upgrade",
]);

const REQUEST_BLOCKED = new Set([
  ...HOP_BY_HOP,
  "authorization",
  "cookie",
  "set-cookie",
  "proxy-connection",
  "host",
  "content-length",
  "x-api-key",
  "api-key",
  "new-api-user",
  "x-winlume-user",
  "x-winlume-user-id",
  "x-winlume-internal-token",
  "x-winlume-internal-user-id",
  "x-winlume-internal-identity",
  "x-winlume-internal-user",
  "forwarded",
  "via",
  "origin",
  "referer",
]);

const RESPONSE_BLOCKED = new Set([
  ...HOP_BY_HOP,
  "set-cookie",
  "content-length",
  "content-encoding",
  // The gateway owns this correlation id; an upstream must not replace it.
  "x-request-id",
]);

function entries(headers: HeaderMap | Headers): Array<[string, string]> {
  const values: Array<[string, string]> = [];
  if (typeof Headers !== "undefined" && headers instanceof Headers) {
    headers.forEach((value, key) => values.push([key.toLowerCase(), value]));
    return values;
  }
  for (const [key, value] of Object.entries(headers as Record<string, string | string[] | undefined>)) {
    if (Array.isArray(value)) {
      for (const item of value) values.push([key.toLowerCase(), item]);
    } else if (typeof value === "string") {
      values.push([key.toLowerCase(), value]);
    }
  }
  return values;
}

function safeValue(value: string): string | undefined {
  if (/[\r\n]/.test(value)) return undefined;
  if (value.length > 16 * 1024) return undefined;
  return value;
}

function connectionScopedHeaders(headers: HeaderMap | Headers): Set<string> {
  return new Set(
    entries(headers)
      .filter(([name]) => name === "connection")
      .flatMap(([, value]) => value.split(","))
      .map((name) => name.trim().toLowerCase())
      .filter(Boolean),
  );
}

function isBlocked(name: string, blocked: Set<string>): boolean {
  if (blocked.has(name)) return true;
  return name.startsWith("x-forwarded-") || name.startsWith("x-winlume-internal-");
}

/** Filter untrusted request headers before forwarding them to an upstream. */
export function filterRequestHeaders(headers: HeaderMap | Headers): Record<string, string> {
  const result: Record<string, string> = {};
  const connectionHeaders = connectionScopedHeaders(headers);
  for (const [name, rawValue] of entries(headers)) {
    if (connectionHeaders.has(name) || isBlocked(name, REQUEST_BLOCKED)) continue;
    const value = safeValue(rawValue);
    if (value === undefined) continue;
    result[name] = result[name] ? `${result[name]}, ${value}` : value;
  }
  return result;
}

/** Filter response headers so upstream transport metadata cannot escape. */
export function filterResponseHeaders(headers: Headers | HeaderMap): Record<string, string> {
  const result: Record<string, string> = {};
  const connectionHeaders = connectionScopedHeaders(headers);
  for (const [name, rawValue] of entries(headers)) {
    if (connectionHeaders.has(name) || isBlocked(name, RESPONSE_BLOCKED)) continue;
    const value = safeValue(rawValue);
    if (value === undefined) continue;
    result[name] = result[name] ? `${result[name]}, ${value}` : value;
  }
  return result;
}

/** Add an internal identity only after server-side authentication. */
export function addTrustedStudioIdentity(
  headers: Record<string, string>,
  userId: string | undefined,
): Record<string, string> {
  if (userId && !/[\r\n]/.test(userId)) headers["new-api-user"] = userId;
  return headers;
}

export function firstHeader(headers: HeaderMap | Headers, name: string): string | undefined {
  return getHeader(headers, name);
}

File: services/gateway/internal/relay/headers.go
package relay

import (
	"net/http"
	"strings"

	"github.com/google/uuid"
)

const maxForwardedHeaderValueBytes = 16 * 1024

var hopByHopHeaders = headerSet(
	"connection",
	"keep-alive",
	"proxy-authenticate",
	"proxy-authorization",
	"te",
	"trailer",
	"transfer-encoding",
	"upgrade",
)

var requestBlockedHeaders = mergedHeaderSet(hopByHopHeaders,
	"authorization",
	"cookie",
	"set-cookie",
	"proxy-connection",
	"host",
	"content-length",
	"accept-encoding",
	"x-api-key",
	"api-key",
	"new-api-user",
	"x-reizo-user",
	"x-reizo-user-id",
	"x-reizo-internal-token",
	"x-reizo-internal-user-id",
	"x-reizo-internal-identity",
	"x-reizo-internal-user",
	"forwarded",
	"via",
	"origin",
	"referer",
)

var responseBlockedHeaders = mergedHeaderSet(hopByHopHeaders,
	"set-cookie",
	"content-length",
	"content-encoding",
	"x-request-id",
)

// RequestHeaderOptions contains only server-controlled header additions.
type RequestHeaderOptions struct {
	Authorization              string
	TrustedUserID              *uuid.UUID
	IncludeNewAPICompatibility bool
}

// FilterRequestHeaders removes caller-controlled credentials, identity,
// forwarding, and connection-scoped values before relay.
func FilterRequestHeaders(source http.Header) http.Header {
	return filterHeaders(source, requestBlockedHeaders)
}

// FilterResponseHeaders removes transport and credential-bearing values that
// must not cross the Gateway response boundary.
func FilterResponseHeaders(source http.Header) http.Header {
	return filterHeaders(source, responseBlockedHeaders)
}

// BuildRequestHeaders filters caller input before adding trusted upstream
// credentials and the optional new-api compatibility identity.
func BuildRequestHeaders(source http.Header, options RequestHeaderOptions) http.Header {
	result := FilterRequestHeaders(source)
	if options.Authorization != "" && safeHeaderValue(options.Authorization) {
		result.Set("Authorization", options.Authorization)
	}
	if options.IncludeNewAPICompatibility && options.TrustedUserID != nil && *options.TrustedUserID != uuid.Nil {
		result.Set("new-api-user", options.TrustedUserID.String())
	}
	return result
}

func filterHeaders(source http.Header, blocked map[string]struct{}) http.Header {
	result := make(http.Header)
	connectionScoped := connectionScopedHeaders(source)
	for name, values := range source {
		normalizedName := strings.ToLower(name)
		if isBlockedHeader(normalizedName, blocked, connectionScoped) {
			continue
		}
		for _, value := range values {
			if safeHeaderValue(value) {
				result.Add(name, value)
			}
		}
	}
	return result
}

func connectionScopedHeaders(headers http.Header) map[string]struct{} {
	result := make(map[string]struct{})
	for headerName, values := range headers {
		if !strings.EqualFold(headerName, "connection") {
			continue
		}
		for _, value := range values {
			for _, name := range strings.Split(value, ",") {
				if name = strings.ToLower(strings.TrimSpace(name)); name != "" {
					result[name] = struct{}{}
				}
			}
		}
	}
	return result
}

func isBlockedHeader(name string, blocked, connectionScoped map[string]struct{}) bool {
	if _, ok := blocked[name]; ok {
		return true
	}
	if _, ok := connectionScoped[name]; ok {
		return true
	}
	return strings.HasPrefix(name, "x-forwarded-") || strings.HasPrefix(name, "x-reizo-internal-")
}

func safeHeaderValue(value string) bool {
	return len(value) <= maxForwardedHeaderValueBytes && !strings.ContainsAny(value, "\r\n")
}

func headerSet(names ...string) map[string]struct{} {
	result := make(map[string]struct{}, len(names))
	for _, name := range names {
		result[name] = struct{}{}
	}
	return result
}

func mergedHeaderSet(base map[string]struct{}, names ...string) map[string]struct{} {
	result := make(map[string]struct{}, len(base)+len(names))
	for name := range base {
		result[name] = struct{}{}
	}
	for _, name := range names {
		result[name] = struct{}{}
	}
	return result
}
